picard/session/track_mover.py

```diff
@@ -0,0 +1,134 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+
+"""Track movement functionality for session management.
+
+This module handles moving files to their designated tracks when loading sessions,
+separating the complex file-to-track movement logic from other concerns.
+"""
+
+from __future__ import annotations
+
+from pathlib import Path
+from typing import Any
+
+from picard.album import Album
+from picard.file import File
+from picard.session.constants import SessionConstants
+from picard.session.retry_helper import RetryHelper
+
+
+class TrackMover:
+    """Handles moving files to their target tracks."""
+
+    def __init__(self, tagger: Any) -> None:
+        """Initialize the track mover.
+
+        Parameters
+        ----------
+        tagger : Any
+            The Picard tagger instance.
+        """
+        self.tagger = tagger
+
+    def move_files_to_tracks(self, album: Album, track_specs: list[tuple[Path, str]]) -> None:
+        """Move files to their designated tracks when ready.
+
+        Parameters
+        ----------
+        album : Album
+            The album containing the tracks.
+        track_specs : list[tuple[Path, str]]
+            List of (file_path, recording_id) tuples to move.
+
+        Notes
+        -----
+        This method schedules file moves when both the file and track are ready.
+        It uses the retry helper to wait for proper conditions.
+        """
+
+        def run_when_album_ready() -> None:
+            for fpath, rid in track_specs:
+                self._schedule_move(fpath, rid, album)
+
+        album.run_when_loaded(run_when_album_ready)
+
+    def _schedule_move(self, fpath: Path, recording_id: str, album: Album) -> None:
+        """Schedule a file move when both file and track are ready.
+
+        Parameters
+        ----------
+        fpath : Path
+            The file path to move.
+        recording_id : str
+            The recording ID of the target track.
+        album : Album
+            The album containing the track.
+        """
+
+        def attempt_move() -> None:
+            file = self.tagger.files.get(str(fpath))
+            if not file or file.state == File.PENDING:
+                return
+
+            rec_to_track = {t.id: t for t in album.tracks}
+            track = rec_to_track.get(recording_id)
+            if track is None:
+                return
+
+            file.move(track)
+
+        def is_ready() -> bool:
+            file = self.tagger.files.get(str(fpath))
+            if not file or file.state == File.PENDING:
+                return False
+
+            rec_to_track = {t.id: t for t in album.tracks}
+            track = rec_to_track.get(recording_id)
+            return track is not None
```

Comment:
I think we can reduce code redundancy a bit here, with something like (totally untested):


```python
    def _get_file_and_track():
        file = self.tagger.files.get(str(fpath))
        if not file or file.state == File.PENDING:
            return None, None

        rec_to_track = {t.id: t for t in album.tracks}
        track = rec_to_track.get(recording_id)
        return file, track

    def attempt_move() -> None:
        file, track = self._get_file_and_track()
        if file is None or track is None:
            return
        file.move(track)

    def is_ready() -> bool:
        file, track = self._get_file_and_track()
        return track is not None
```